drivers/clock_control/clock_control_nrf2_hsfll.c

```diff
@@ -0,0 +1,211 @@
+	enum dvfs_frequency_setting new_setting;
+	int rc;
+
+	active_options = clock_config_update_begin(work);
+	if (active_options & BIT(2)) {
+		new_setting = DVFS_FREQ_HIGH;
+	} else if (active_options & BIT(1)) {
+		new_setting = DVFS_FREQ_MEDLOW;
+	} else {
+		new_setting = DVFS_FREQ_LOW;
+	}
+
+	rc = dvfs_service_handler_change_freq_setting(new_setting);
+	if (rc < 0) {
+		clock_config_update_end(&dev_data->clk_cfg, rc);
+	} else {
+		k_timer_start(&dev_data->timer, UPDATE_TIMEOUT_US, K_NO_WAIT);
+	}
+}
+
+static struct onoff_manager *hsfll_find_mgr(const struct device *dev,
+					    const struct nrf_clock_spec *spec)
+{
+	struct hsfll_dev_data *dev_data = dev->data;
+	uint32_t frequency;
+
+	if (!spec) {
+		return &dev_data->clk_cfg.onoff[0].mgr;
+	}
+
+	if (spec->accuracy || spec->precision) {
+		LOG_ERR("invalid specification of accuracy or precision");
+		return NULL;
+	}
+
+	frequency = spec->frequency == NRF_CLOCK_CONTROL_FREQUENCY_MAX
+		  ? max_frequency
+		  : spec->frequency;
+
+	for (int i = 0; i < ARRAY_SIZE(oppoint_freq); ++i) {
+		if (frequency > oppoint_freq[i]) {
+			continue;
+		}
+
+		return &dev_data->clk_cfg.onoff[i].mgr;
+	}
+
+	LOG_ERR("invalid frequency");
+	return NULL;
+}
+#endif /* CONFIG_NRFS_HAS_DVFS_SERVICE */
+
+static int api_request_hsfll(const struct device *dev,
+			     const struct nrf_clock_spec *spec,
+			     struct onoff_client *cli)
+{
+#ifdef CONFIG_NRFS_HAS_DVFS_SERVICE
+	struct onoff_manager *mgr = hsfll_find_mgr(dev, spec);
+
+	if (mgr) {
+		return onoff_request(mgr, cli);
+	}
+
+	return -EINVAL;
+#else
+	return -ENOTSUP;
+#endif
+}
+
+static int api_release_hsfll(const struct device *dev,
+			     const struct nrf_clock_spec *spec)
+{
+#ifdef CONFIG_NRFS_HAS_DVFS_SERVICE
+	struct onoff_manager *mgr = hsfll_find_mgr(dev, spec);
+
+	if (mgr) {
+		return onoff_release(mgr);
+	}
+
+	return -EINVAL;
+#else
+	return -ENOTSUP;
+#endif
+}
+
+static int api_cancel_or_release_hsfll(const struct device *dev,
+				       const struct nrf_clock_spec *spec,
+				       struct onoff_client *cli)
+{
+#ifdef CONFIG_NRFS_HAS_DVFS_SERVICE
+	struct onoff_manager *mgr = hsfll_find_mgr(dev, spec);
+
+	if (mgr) {
+		return onoff_cancel_or_release(mgr, cli);
+	}
+
+	return -EINVAL;
+#else
+	return -ENOTSUP;
+#endif
+}
+
+static int api_get_rate_hsfll(const struct device *dev,
+			      clock_control_subsys_t sys,
+			      uint32_t *rate)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sys);
+
+	*rate = nrf_hsfll_clkctrl_mult_get(NRF_HSFLL) * MHZ(16);
+
+	return 0;
+}
+
+static int hsfll_init(const struct device *dev)
+{
+#ifdef CONFIG_NRFS_HAS_DVFS_SERVICE
+	struct hsfll_dev_data *dev_data = dev->data;
+	int rc;
+
+	rc = clock_config_init(&dev_data->clk_cfg,
+			       ARRAY_SIZE(dev_data->clk_cfg.onoff),
+			       hsfll_work_handler);
+	if (rc < 0) {
+		return rc;
+	}
+
+	k_timer_init(&dev_data->timer, hsfll_update_timeout_handler, NULL);
+
+	dvfs_service_handler_register_freq_setting_applied_callback(
+		freq_setting_applied_cb);
```

Comment:
I expected to find requesting the initial frequency (the lowest one?) somewhere around here. Is it in some other place I missed?

Reply:
At least Secure Domain should set the lowest oppoint during initialisation. Without this, dvfs scaling on App core will not work.

Reply:
Enabling `CONFIG_NRFS_LOCAL_DOMAIN_DVFS_SCALE_DOWN_AFTER_INIT` should do the job: https://github.com/zephyrproject-rtos/zephyr/blob/5d87cd654fbe2d430bcdaf39c666683abe65f290/modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c#L291-L294
I'll add this as default.

Reply:
Now I see a race condition: the `ld_dvfs_something` module sets the initial frequency when ABB gets applied (max). In the meantime, this module is expected to set the initial frequency (probably min, or user-defined). These operations can happen in any order resulting in unknown initial frequency.

In the current design with `ld_dvfs_handler` I have a feeling that the initial frequency should be requested from clock_control to ld_dvfs_handler. The `ld_dvfs_handler` should apply the requested initial frequency immediately (if ABB has already been applied) or after ABB gets applied.

Reply:
Looks good. I've just spotted a potential race condition in ld_dvfs_handler.c:

The requested frequency is correctly set by the procedure you linked. However it might get incorrectly overwritten by `static void dvfs_service_handler_set_initial_hsfll_config(void)`. Is my observation correct?

Reply:
From what I can tell, `dvfs_service_handler_task()` waits for `dvfs_service_sync_sem` before calling `dvfs_service_handler_change_freq_setting()`, so `dvfs_service_handler_set_initial_hsfll_config()` needs to complete first. Unless `NRFS_DVFS_EVT_INIT_DONE` is received twice (what would mean that something is really off), it should be fine.

Reply:
Right. I didn't realize that the only invocations of these functions are protected from races with a semaphore. It's not the cleanest code (adding more calls to these functions can cause race conditions), but now it seems fine.

Reply:
CONFIG_NRFS_LOCAL_DOMAIN_DVFS_SCALE_DOWN_AFTER_INIT this config was introduced to enable option to change DVFS settings from application core without other existing solution to do that. I don't know if this shouldn't be explicitly called by secdom for example when all initialization is done?